tools/gbcpal.c

```diff
@@ -0,0 +1,156 @@
+#define PROGRAM_NAME "gbcpal"
+#define USAGE_OPTS "[-h|--help] [-r|--reverse] out.gbcpal [in.gbcpal...]"
+
+#include "common.h"
+
+bool reverse;
+
+void parse_args(int argc, char *argv[]) {
+	struct option long_options[] = {
+		{"reverse", no_argument, 0, 'r'},
+		{"help", no_argument, 0, 'h'},
+		{0}
+	};
+	for (int opt; (opt = getopt_long(argc, argv, "rh", long_options)) != -1;) {
+		switch (opt) {
+		case 'r':
+			reverse = true;
+			break;
+		case 'h':
+			usage_exit(0);
+			break;
+		default:
+			usage_exit(1);
+		}
+	}
+}
+
+struct Color {
+	uint8_t r, g, b;
+};
+
+uint16_t pack_color(struct Color color) {
+	return (color.b << 10) | (color.g << 5) | color.r;
+}
+
+struct Color unpack_color(uint8_t lo, uint8_t hi) {
+	return (struct Color){
+		.r = lo & 0x1f,
+		.g = ((hi << 3) & 0x18) | ((lo >> 5) & 0x07),
+		.b = (hi >> 2) & 0x1f,
+	};
+}
+
+double luminance(struct Color color) {
+	return 0.299 * color.r * color.r + 0.587 * color.g * color.g + 0.114 * color.b * color.b;
```

Comment:
This is the standard conversion from RGB to Y (luminance), used by jpeg encoding and probably lots of other things.
I use the exact same coefficients [here](https://github.com/dannye/jed/blob/0250dabe87698c216dc906ae7213cc319dddbdd5/src/encoder.cpp#L107C12-L107C98), for example.

Reply:
What breakup do you have in mind? It's the same function I've used in a few Python helper scripts:

```python
def luminance(c):
	r, g, b = c
	return 0.299 * r**2 + 0.587 * g**2 + 0.114 * b**2
```

Reply:
Oh, actually I didn't notice at first that you are squaring each color component. That is not the standard luminance conversion used by jpeg etc. But it's probably the standard somewhere! I've just never seen it before. I'd be curious to hear the benefits.

Reply:
I've used it for many years, quoting from some source that defined luminance as the sqrt of sum of squares (i.e. a Euclidean distance) but omitting the sqrt when it's unnecessary (e.g. for comparisons).

Google leads to https://stackoverflow.com/questions/596216/formula-to-determine-perceived-brightness-of-rgb-color leads to https://alienryderflex.com/hsp.html which looks familiar (and IMO based on the StackOverflow illustrations is the best choice of the given options).

Reply:
Unsurprisingly, the choice of luminance formula *does* affect which sprites count as reversed.

These sprites need `--reverse` in either case: `farfetch_d`, `fearow`, `jynx`, `porygon2`, `scyther`, `spearow`, `swimmer_m`

These sprites need `--reverse` with the current squaring formula, but not with a linear one: `caterpie`, `diglett`, `dugtrio`, `magnemite`, `magneton`, `marill`, `shellder`

These sprites would need `--reverse` with a linear formula, but not with the current one: `hitmonlee`, `porygon`

Whichever formula we go with, I'd need to update [palfix.py](https://github.com/pret/gb-asm-tools/blob/master/tools/palfix.py) and [revpals.py](https://github.com/pret/gb-asm-tools/blob/master/tools/revpals.py) to agree with it.

Reply:
Because I just cannot let go of this problem:

![image](https://github.com/user-attachments/assets/67a00a4d-da10-44c9-b93d-027a851f3da8)

- The shortest code is `0.299 * r + 0.587 * g + 0.114 * b`, which affects the fewest 9 sprites.
- The most theoretically accurate code is `0.2126 * r**2.2 + 0.7152 * g**2.2 + 0.0722 * b**2.2`, which still affects just 10 sprites.

The other options, including my usual HSP `0.299 * r**2 + 0.587 * g**2 + 0.114 * b**2`, compromise both brevity and accuracy, and all end up affecting more sprites anyway.

Reply:
For reference, here are the sprites which at least one algorithm considers to be reversed:
![image](https://github.com/user-attachments/assets/e544f0f0-ffe9-43e3-99e1-6abc1192227f)

And here they are with SameBoy's "Modern - Reduce Contrast" correction (via [gbc_colors.py](https://github.com/Rangi42/polishedcrystal/blob/9bit/utils/gbc_colors.py)):
![image](https://github.com/user-attachments/assets/bb7c496b-277b-44cb-a0e7-166dd2bcd3b8)